packages/data-viz/src/core/HeatmapChart/index.tsx

```diff
@@ -18,101 +19,176 @@ export interface HeatmapChartProps
-   */
 }
 
-function HeatmapChart(
-  props: HeatmapChartProps,
-  ref: ForwardedRef<HTMLDivElement>
-): JSX.Element {
-  const {
-    width,
-    height,
-    echartsRendererMode = "canvas",
-    onChartMouseMove,
-    xAxisData,
-    yAxisData,
-    data,
-    encode,
-    itemStyle,
-    symbolSize = 5,
-    grid,
-    ...rest
-  } = props;
-
-  if (!width || !height) {
-    throw Error("Chart must have width and height > 0");
-  }
+const HeatmapChart = forwardRef(
+  (
+    props: HeatmapChartProps,
+    ref: ForwardedRef<HTMLDivElement>
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+  ): JSX.Element => {
+    const {
+      axisPointer,
+      width,
+      height,
+      echartsRendererMode = "svg",
+      camera,
+      onEvents,
+      xAxisData,
+      yAxisData,
+      data,
+      encode,
+      emphasis,
+      itemStyle,
+      symbol,
+      symbolSize = 5,
+      grid,
+      options,
+      ...rest
+    } = props;
+
+    // Validate width and height
+    if (!width || !height) {
+      throw Error("Chart must have width and height > 0");
+    }
 
-  const [isChartInitialized, setIsChartInitialized] = useState(false);
+    // State for chart initialization
+    const [isChartInitialized, setIsChartInitialized] = useState(false);
+
+    // State for ECharts instance
+    const [chart, setChart] = useState<ECharts | null>(null);
+
+    // Ref for the chart container
+    const innerRef = useRef<HTMLDivElement | null>(null);
+
+    // Function to initialize the chart
+    const initChart = useCallback(() => {
+      const { current } = innerRef;
+
+      if (
+        !current ||
+        isChartInitialized ||
+        // (thuang): echart's `init()` will throw error if the container has 0 width or height
+        current?.getAttribute("height") === "0" ||
+        current?.getAttribute("width") === "0"
+      ) {
+        return;
+      }
+
+      // Initialize ECharts instance
+      const rawChart = init(current, EMPTY_OBJECT, {
+        renderer: echartsRendererMode,
+        useDirtyRect: true,
+      });
 
-  const [chart, setChart] = useState<ECharts | null>(null);
-  const innerRef = useRef<HTMLDivElement | null>(null);
+      setIsChartInitialized(true);
+
+      // Bind events if provided
+      if (onEvents) {
+        bindEvents(rawChart, onEvents);
+      }
+
+      setChart(rawChart);
+
+      // Cleanup function
+      return () => {
+        chart?.dispose();
+
+        // Unbind events if provided
+        if (onEvents) {
+          for (const eventName in onEvents) {
+            if (
+              Object.prototype.hasOwnProperty.call(onEvents, eventName) &&
+              typeof eventName === "string" &&
+              typeof onEvents[eventName] === "function"
+            ) {
+              rawChart.off(eventName, onEvents[eventName]);
+            }
+          }
+        }
+      };
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
what's the reason for disabling this rule here? Thanks!

Reply:
Aah, this one was redundant and I removed it! 😬